internal/policy/verify.go

```diff
@@ -0,0 +1,140 @@
+package policy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/adityasaky/gittuf/internal/gitinterface"
+	"github.com/adityasaky/gittuf/internal/rsl"
+	"github.com/adityasaky/gittuf/internal/tuf"
+	"github.com/go-git/go-git/v5"
+)
+
+var (
+	ErrUnauthorizedSignature = errors.New("unauthorized signature")
+)
+
+func VerifyRef(ctx context.Context, repo *git.Repository, target string) error {
+	// 1. Trace RSL back to the start
+	firstEntryTmp, err := rsl.GetFirstEntry(repo)
+	if err != nil {
+		return err
+	}
+	firstEntry := firstEntryTmp.(*rsl.Entry)
+
+	// 2. Find latest entry for target
+	latestEntryTmp, err := rsl.GetLatestEntryForRef(repo, target)
+	if err != nil {
+		return err
+	}
+	latestEntry := latestEntryTmp.(*rsl.Entry)
+
+	// 3. Do a relative verify from start entry to the latest entry
+	return VerifyRelativeForRef(ctx, repo, firstEntry, latestEntry, target)
+}
+
+func VerifyRelativeForRef(ctx context.Context, repo *git.Repository, firstEntry *rsl.Entry, lastEntry *rsl.Entry, target string) error {
+	entryStack := []*rsl.Entry{lastEntry}
+
+	var currentPolicy *State
+	// 1. Identify policy applicable at firstEntry
+	policyEntry, err := rsl.GetLatestEntryForRefBefore(repo, PolicyRef, firstEntry.ID)
+	if err != nil {
+		return err
+	}
+	state, err := LoadStateForEntry(ctx, repo, policyEntry)
+	if err != nil {
+		return err
+	}
+	currentPolicy = state
+
+	// 2. Enumerate RSL entries between firstEntry and lastEntry, ignoring irrelevant ones
+	iteratorEntry := lastEntry
+	for {
+		if iteratorEntry.GetID() == firstEntry.ID {
+			break
+		}
+
+		parentEntryTmp, err := rsl.GetParentForEntry(repo, iteratorEntry)
+		if err != nil {
+			return err
+		}
+		parentEntry := parentEntryTmp.(*rsl.Entry) // TODO: handle annotations
```

Comment:
Type switching is usually a signal that the interface may not be what you want - e.g. why have `GetParentForEntry` take in a `EntryType` instead of a `*Entry`? 🤔

Reply:
I've swapped it to be *Entry but we'll still have to handle annotations. Right now we don't create any annotations but in future, we'd also be tracking them here instead of always assuming it's *Entry. Then, we'd check the annotations to see if other entries are to be skipped.